File: src/modules/email.ts
import { HttpClient, toRecipientString } from '../http-client';
import {
  EmailHistoryResult,
  EmailMessage,
  HttpMethod,
  MsGineValidationError,
  PaginationOptions,
  SendEmailOptions,
} from '../types';

export class EmailModule {
  constructor(private readonly http: HttpClient) {}

  /**
   * Send an email.
   *
   * At least one of `text` or `html` is required.
   *
   * @example
   * ```ts
   * const msg = await client.email.send({
   *   from: 'no-reply@mail.msgine.net',
   *   to: 'user@example.com',
   *   subject: 'Welcome to MsGine!',
   *   html: '<h1>Hello!</h1>',
   * });
   * ```
   */
  async send(options: SendEmailOptions): Promise<EmailMessage> {
    validate(options);

    return this.http.request<EmailMessage>({
      method: HttpMethod.POST,
      path: '/developers/email',
      body: {
        from: options.from,
        to: toRecipientString(options.to),
        cc: options.cc ? toRecipientString(options.cc) : undefined,
        bcc: options.bcc ? toRecipientString(options.bcc) : undefined,
        subject: options.subject,
        text: options.text,
        html: options.html,
        scheduledFor: options.scheduledFor,
        idempotencyKey: options.idempotencyKey,
      },
    });
  }

  /**
   * Get paginated email send history.
   *
   * @example
   * ```ts
   * const { items } = await client.email.history({ limit: 50 });
   * ```
   */
  async history(options: PaginationOptions = {}): Promise<EmailHistoryResult> {
    const raw = await this.http.request<{
      messages: EmailMessage[];
      total: number;
      page: number;
      pages: number;
    }>({
      method: HttpMethod.GET,
      path: '/developers/email/history',
      queryParams: { page: options.page ?? 1, limit: options.limit ?? 20 },
    });

    return { items: raw.messages, total: raw.total, page: raw.page, pages: raw.pages };
  }
}

// ── Validation ────────────────────────────────────────────────────────────────

function validate(o: SendEmailOptions): void {
  if (!o.from || o.from.trim().length === 0)
    throw new MsGineValidationError('from is required', 'from');
  const to = Array.isArray(o.to) ? o.to : o.to.split(',');
  if (to.length === 0 || to.some((e) => e.trim().length === 0))
    throw new MsGineValidationError('to: at least one recipient is required', 'to');
  if (!o.subject || o.subject.trim().length === 0)
    throw new MsGineValidationError('subject is required', 'subject');
  if (o.subject.length > 200)
    throw new MsGineValidationError('subject must be 200 characters or fewer', 'subject');
  if (!o.text && !o.html)
    throw new MsGineValidationError('at least one of text or html is required');
}

File: go/email.go
package msgine

import (
	"context"
	"strings"
)

// EmailService provides methods for sending and listing email messages.
type EmailService struct {
	t *transport
}

type sendEmailRequest struct {
	From           string `json:"from"`
	To             string `json:"to"`
	Subject        string `json:"subject"`
	HTML           string `json:"html,omitempty"`
	Text           string `json:"text,omitempty"`
	CC             string `json:"cc,omitempty"`
	BCC            string `json:"bcc,omitempty"`
	ScheduledFor   string `json:"scheduledFor,omitempty"`
	IdempotencyKey string `json:"idempotencyKey,omitempty"`
}

// Send sends an email. At least one of HTML or Text is required.
//
// To, CC, and BCC accept either a single address or a comma-separated string.
// Use the helper JoinAddresses if you have a []string.
func (s *EmailService) Send(ctx context.Context, opts SendEmailOptions) (*EmailMessage, error) {
	if opts.From == "" {
		return nil, &ValidationError{Field: "from", Message: "from is required"}
	}
	if len(opts.To) == 0 {
		return nil, &ValidationError{Field: "to", Message: "at least one recipient is required"}
	}
	if opts.Subject == "" {
		return nil, &ValidationError{Field: "subject", Message: "subject is required"}
	}
	if len(opts.Subject) > 200 {
		return nil, &ValidationError{Field: "subject", Message: "subject must be 200 characters or fewer"}
	}
	if opts.HTML == "" && opts.Text == "" {
		return nil, &ValidationError{Message: "at least one of HTML or Text is required"}
	}

	body := sendEmailRequest{
		From:           opts.From,
		To:             strings.Join(opts.To, ","),
		Subject:        opts.Subject,
		HTML:           opts.HTML,
		Text:           opts.Text,
		ScheduledFor:   opts.ScheduledFor,
		IdempotencyKey: opts.IdempotencyKey,
	}
	if opts.CC != nil {
		body.CC = strings.Join(opts.CC, ",")
	}
	if opts.BCC != nil {
		body.BCC = strings.Join(opts.BCC, ",")
	}

	var result EmailMessage
	if err := s.t.do(ctx, "POST", "/developers/email", nil, body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// History returns a paginated list of sent emails.
func (s *EmailService) History(ctx context.Context, page, limit int) (*PaginatedResult[EmailMessage], error) {
	q := queryValues("page", page, "limit", limit)
	var result PaginatedResult[EmailMessage]
	if err := s.t.do(ctx, "GET", "/developers/email/history", q, nil, &result); err != nil {
		return nil, err
	}
	return &result, nil
}
